### PR TITLE
[Torch] PTQ conformance test

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

`quantize_impl` function for Torch PTQ
Torch PTQ is enabled in the conformance test.
Additional parameter `--backends` allows to choose between PTQ backends (`TORCH`,`TORCH_PTQ`,`ONNX`,`OV_NATIVE`,`OV`)
Additional columns: `FPS diff` and `top 1 diff` was added for each quantization backend + aggregated `status` column was added 

### Reason for changes

To enable Torch PTQ in conformance tests
To make it possible to pick PTQ backend in the conformance test

### Related tickets

98133

### CODE REVIEW

File: tests/post_training/test_quantize_conformance.py [hunk 1]
```diff
@@ -234,11 +237,175 @@ def quantize_ov_native(model: ov.Model,
...
 
-def run_ptq_timm(data, output, model_name, model_quantization_params, process_connection): # pylint: disable=W0703
+def quantize_torch_ptq(model: torch.nn.Module,
+                       calibration_dataset: nncf.Dataset,
+                       preset: nncf.QuantizationPreset = nncf.QuantizationPreset.PERFORMANCE,
+                       target_device: nncf.TargetDevice = nncf.TargetDevice.ANY,
+                       subset_size: int = 300,
+                       fast_bias_correction: bool = True,
+                       model_type: Optional[nncf.ModelType] = None,
+                       ignored_scope: Optional[nncf.IgnoredScope] = None) -> NNCFNetwork:
```
Comment (KodiaqQ) [inline tests/post_training/test_quantize_conformance.py:247]: `pt_impl_experimental` returns `torch.nn.Module`, not `NNCFNetwork`.

File: nncf/torch/quantization/quantize.py [hunk 2]
```diff
@@ -213,3 +218,45 @@ def send_to_device(tensor):
...
+        preset: QuantizationPreset,
+        target_device: TargetDevice,
+        subset_size: int,
+        fast_bias_correction: bool,
+        model_type: Optional[ModelType] = None,
+        ignored_scope: Optional[IgnoredScope] = None) -> torch.nn.Module:
+    """
+    Experimental implementation of the `quantize()` method for the PyTorch backend.
+    """
+    if fast_bias_correction is False:
```
Comment (KodiaqQ) [inline nncf/torch/quantization/quantize.py:235]: Neither is the `Fast-` nor that just `BiasCorrection` supported for now for the PyTorch PTQ, isn't it?

File: tests/post_training/test_quantize_conformance.py [hunk 3]
```diff
@@ -261,142 +429,25 @@ def transform_fn(data_item):
...
+
+        process_connection.send(runinfos)
     except Exception as error:
         traceback_path = Path.joinpath(output_folder, model_name + '_error_log.txt')
         create_error_log(traceback_path)
-        error_message = f'An error occurred while running {model_name}. Traceback: {traceback_path}'
-        result = [model_name, error_message, '-', '-', '-', '-', '-', '-', '-', '-', '-']
-        process_connection.send(result)
+        status = f'An error occurred while running {model_name}. Traceback: {traceback_path}'
+        runinfos[QuantizationBackend.FP32] = RunInfo(-1, -1, status)
```
Comment (KodiaqQ) [inline tests/post_training/test_quantize_conformance.py:449]: This `except` section was responsible for the whole run of the concrete model, not the concrete backend. So, the message about the error should be widespread to the whole list of backends, isn't it?

**Comment by KodiaqQ** [comment]
It would be great to have a green build with the changes and first `PyTorch` results.

**Comment by daniil-lyakhov** [comment]
> It would be great to have a green build with the changes and first `PyTorch` results.

Build 38 is running

File: tests/post_training/test_quantize_conformance.py (hunk 3 above)
Comment (daniil-lyakhov) [reply tests/post_training/test_quantize_conformance.py:449]: Yes, and it actually will be widespread, I move this logic over here: https://github.com/openvinotoolkit/nncf/pull/1591/files#diff-a6044884bf87069635cb96de0928ebd1c2d0a1a849693480684139737e35420fR64-R73

So, if `runinfos` has no info about  backends, error message only for FP32 model appears in this row and `-` appears in backend columns

File: tests/post_training/test_quantize_conformance.py (hunk 1 above)
Comment (daniil-lyakhov) [reply tests/post_training/test_quantize_conformance.py:247]: Actually PTQ uses `NNCFNetwork` to retrieve `NNCFGraph` and to add quantization modules. `NNCFNetwork`  is returned from `pt_impl_experimental` as well 
image

File: nncf/torch/quantization/quantize.py (hunk 2 above)
Comment (daniil-lyakhov) [reply nncf/torch/quantization/quantize.py:238]: Yes, I've added warning to make it clearer

File: tests/post_training/test_quantize_conformance.py (hunk 1 above)
Comment (KodiaqQ) [reply tests/post_training/test_quantize_conformance.py:247]: Okay, I see. But then, it means that the output type hint for `quantize_impl_experimental` is not correct? It's `torch.nn.Module`, not `NNCFNetwork`.

File: nncf/torch/quantization/quantize.py (hunk 2 above)
Comment (KodiaqQ) [reply nncf/torch/quantization/quantize.py:238]: Let's add it for `quantize_impl_experimental` as well.

Comment (daniil-lyakhov) [reply nncf/torch/quantization/quantize.py:238]: Sure, done!

File: tests/post_training/test_quantize_conformance.py (hunk 1 above)
Comment (daniil-lyakhov) [reply tests/post_training/test_quantize_conformance.py:247]: You are right. BTW it's fair for TORCH backend as well

Comment (daniil-lyakhov) [reply tests/post_training/test_quantize_conformance.py:247]: Done

File: nncf/torch/quantization/quantize.py [hunk 4]
```diff
@@ -148,13 +151,15 @@ def quantize_impl(model: torch.nn.Module,
...
                   model_type: Optional[ModelType] = None,
-                  ignored_scope: Optional[IgnoredScope] = None) -> torch.nn.Module:
+                  ignored_scope: Optional[IgnoredScope] = None) -> NNCFNetwork:
     """
     Implementation of the `quantize()` method for the PyTorch backend.
     """
     if fast_bias_correction is False:
         raise ValueError(f'fast_bias_correction={fast_bias_correction} is not '
                           'supported')
+    nncf_logger.warning('Bias correction and fast bias correction algorithms'
```
Comment (alexsu52) [inline nncf/torch/quantization/quantize.py:161]: Please revert these changes. Torch supports default behavior and raise error if fast_bias_correction is False. This warning just generate noise.

File: nncf/quantization/algorithms/post_training/algorithm.py [hunk 5]
```diff
@@ -191,6 +191,8 @@ def _apply(self,
             # TODO (l-bat): Remove after OpenVINO Native is removed from experimental
             if backend == BackendType.OPENVINO:
                 nncf_logger.warning('You are using experimental OpenVINO backend for the Post-training quantization.')
+            if backend == BackendType.TORCH:
+                nncf_logger.warning('You are using experimental Torch backend for the Post-training quantization.')
```
Comment (alexsu52) [inline nncf/quantization/algorithms/post_training/algorithm.py:195]: I suggest removing it because when we switch `quantize()` to Torch implementation of the post-training quantization algorithm, it will not be an experimental implementation.

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/quantization/algorithms/post_training/algorithm.py (hunk 5 above)
Comment (KodiaqQ) [reply nncf/quantization/algorithms/post_training/algorithm.py:195]: Agree

File: nncf/torch/quantization/quantize.py [hunk 6]
```diff
@@ -148,13 +151,15 @@ def quantize_impl(model: torch.nn.Module,
                   subset_size: int,
                   fast_bias_correction: bool,
                   model_type: Optional[ModelType] = None,
-                  ignored_scope: Optional[IgnoredScope] = None) -> torch.nn.Module:
+                  ignored_scope: Optional[IgnoredScope] = None) -> NNCFNetwork:
```
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply nncf/torch/quantization/quantize.py:154]: Agree with @alexsu52

File: nncf/torch/quantization/quantize.py (hunk 4 above)
Comment (KodiaqQ) [reply nncf/torch/quantization/quantize.py:161]: Agree

File: nncf/quantization/algorithms/post_training/algorithm.py (hunk 5 above)
Comment (daniil-lyakhov) [reply nncf/quantization/algorithms/post_training/algorithm.py]: Done

File: nncf/torch/quantization/quantize.py (hunk 4 above)
Comment (daniil-lyakhov) [reply nncf/torch/quantization/quantize.py]: Done

**Comment by daniil-lyakhov** [comment]
Example of generated table (converted from .csv to markdown)

| # Model name | FP32 top 1          | Torch INT8 top 1    | Torch PTQ INT8 top 1 | ONNX INT8 top 1     | OV Native INT8 top 1 | Openvino INT8 top 1 | FP32 FPS | Torch INT8 FPS | Torch PTQ INT8 FPS | ONNX INT8 FPS | OV Native INT8 FPS | Openvino INT8 FPS | Torch INT8 top 1 diff | Torch PTQ INT8 top 1 diff | ONNX INT8 top 1 diff | OV Native INT8 top 1 diff | Openvino INT8 top 1 diff | Torch INT8 FPS speedup | Torch PTQ INT8 FPS speedup | ONNX INT8 FPS speedup | OV Native INT8 FPS speedup | Openvino INT8 FPS speedup | Status |
|--------------|---------------------|---------------------|----------------------|---------------------|----------------------|---------------------|----------|----------------|--------------------|---------------|--------------------|-------------------|-----------------------|---------------------------|----------------------|---------------------------|--------------------------|------------------------|----------------------------|-----------------------|----------------------------|---------------------------|--------|
| vgg11        | 0.09248407643312102 | 0.09248407643312102 | 0.09248407643312102  | 0.09248407643312102 | 0.09248407643312102  | 0.09222929936305732 | 116.25   | 473.45         | 446.78             | 453.39        | 483.22             | 470.47            | 0.0                   | 0.0                       | 0.0                  | 0.0                       | 0.00025477707006370254   | 4.07268817204301       | 3.843268817204301          | 3.9001290322580644    | 4.156731182795699          | 4.047053763440861         |        |

**Review by KodiaqQ — APPROVED** [review]
